### ISSUE TITLE
Division by zero crash

### ISSUE THREAD

**Issue by Google-Autofuzz** [uclouvain/openjpeg#937]
Hello OpenJpeg,

As part of our fuzzing efforts at Google, we have identified an issue in version 8650b70e064 (from head).  To reproduce requires compiling the project with the LLVM compiler, taking advantage of the sanitizers that it offers (this issue was discovered using AddressSanitizer).
 
To reproduce you will need to build your project using that sanitizer, and execute the attached stub code on the reproducer input that we have also provided.  This stub code could also serve as a useful template for fuzzing in your project with libFuzzer and/or AFL, which may help you uncover additional issues.  Some documentation on how to get started with libFuzzer is here:
Getting Started Documentation
LibFuzzer Tutorial 
OSS-Fuzz target example
 
The following options / environment variables may be necessary for accurate reproduction of the issue as well:
ASAN_OPTIONS="exitcode=1,handle_segv=1,detect_leaks=1,leak_check_at_exit=1,allocator_may_return_null=1,detect_odr_violation=0"
MSAN_OPTIONS=...
 
The sanitizer error that we encountered is here:
third_party/openjpeg2/src/lib/openjp2/pi.c:523:19: runtime error: division by zero
    #0  in opj_pi_next_cprl 
    #1  in opj_t2_decode_packets 
    #2  in opj_tcd_t2_decode
    #3  in opj_tcd_decode_tile
    #4  in opj_j2k_decode_tile 
    #5  in opj_j2k_decode_tiles 
    #6  in opj_j2k_exec 
    #7  in opj_j2k_decode
    #8  in DecodeAs(CODEC_FORMAT, unsigned char const*, unsigned long) 
    #9  in LLVMFuzzerTestOneInput
    #10  in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) 
    #11  in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long) 
    #12  in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) 
    #13  in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) 
    #14  in main 
    #15  in __libc_start_main
    #16  in _start 
 
Other relevant info/repro instructions: After building with ASAN, this should trigger the bug (note .j2k extension must be added): ./opj_decompress -i /poc-752c5d6632455bac7732b47da0ba98eacf4cfd9bd98b53f5a703504a8ffb6cd5.j2k -o foo.pgm

# ./opj_decompress -i /fuzzing/og-b38087388.j2k -o foo.pgm
[INFO] Start to read j2k main header (0).
[WARNING] Unknown marker
[INFO] Main header has been correctly decoded.
[INFO] No decoded area parameters, set the decoded area to the whole image
[INFO] Psot value of the current tile-part is equal to zero, we assuming it is the last tile-part of the codestream.
[INFO] Header of tile 1 / 1 has been read.
/fuzzing/openjpeg/src/lib/openjp2/pi.c:523:19: runtime error: division by zero
[poc-752c5d6632455bac7732b47da0ba98eacf4cfd9bd98b53f5a703504a8ffb6cd5.zip](https://github.com/uclouvain/openjpeg/files/1023275/poc-752c5d6632455bac7732b47da0ba98eacf4cfd9bd98b53f5a703504a8ffb6cd5.zip)

 
We will gladly work with you so you can successfully confirm and reproduce this issue. Do let us know if you have any feedback surrounding the documentation.  Once you have reproduced the issue, we’d appreciate to learn your expected timeline for an update to be released. With any fix, please attribute the report to “Google Autofuzz project”.
 
We are also pleased to inform you that your project is eligible for inclusion to the OSS-Fuzz project, which can provide additional continuous fuzzing, and encourage you to investigate integration options.
 
Don’t hesitate to let us know if you have any questions!
 
Google AutoFuzz Team

(38087388)

**Comment by rwhitworth** [comment]
Valgrind output from the same error found on a different fuzzing effort:
```
==3880675== Using Valgrind-3.10.0 and LibVEX; rerun with -h for copyright info
==3880675== Command: ./opj_decompress -i output-decompress/jpeg-1/crashes/id:000063,sig:08,src:001500,op:flip4,pos:49 -o /tmp/output.tif
==3880675==

===========================================
The extension of this file is incorrect.
FOUND s:49. SHOULD BE .j2k or .jpc or .j2c
===========================================

[INFO] Start to read j2k main header (0).
[INFO] Main header has been correctly decoded.
[INFO] No decoded area parameters, set the decoded area to the whole image
[INFO] Psot value of the current tile-part is equal to zero, we assuming it is the last tile-part of the codestream.
[INFO] Header of tile 1 / 1 has been read.
==3880675==
==3880675== Process terminating with default action of signal 8 (SIGFPE)
==3880675==  Integer divide by zero at address 0x803FAC897
==3880675==    at 0x4EA22C1: opj_pi_next_cprl (pi.c:564)
==3880675==    by 0x4EA22C1: opj_pi_next (pi.c:1938)
==3880675==    by 0x4ED027F: opj_t2_decode_packets (t2.c:424)
==3880675==    by 0x4EDA616: opj_tcd_t2_decode (tcd.c:1636)
==3880675==    by 0x4EDA616: opj_tcd_decode_tile (tcd.c:1386)
==3880675==    by 0x4E5EFD0: opj_j2k_decode_tile (j2k.c:8604)
==3880675==    by 0x4E76DDA: opj_j2k_decode_tiles (j2k.c:10303)
==3880675==    by 0x4E5B09D: opj_j2k_exec (j2k.c:7783)
==3880675==    by 0x4E63055: opj_j2k_decode (j2k.c:10522)
==3880675==    by 0x4E966B4: opj_decode (openjpeg.c:441)
==3880675==    by 0x406312: main (opj_decompress.c:1435)
==3880675==
==3880675== HEAP SUMMARY:
==3880675==     in use at exit: 1,224,161 bytes in 187 blocks
==3880675==   total heap usage: 191 allocs, 4 frees, 1,226,369 bytes allocated
==3880675==
==3880675== LEAK SUMMARY:
==3880675==    definitely lost: 0 bytes in 0 blocks
==3880675==    indirectly lost: 0 bytes in 0 blocks
==3880675==      possibly lost: 0 bytes in 0 blocks
==3880675==    still reachable: 1,224,161 bytes in 187 blocks
==3880675==         suppressed: 0 bytes in 0 blocks
==3880675== Rerun with --leak-check=full to see details of leaked memory
==3880675==
==3880675== For counts of detected and suppressed errors, rerun with: -v
==3880675== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Floating point exception
```

**Comment by rwhitworth** [comment]
Is this a duplicate of #855?

**Comment by szukw000** [comment]
@Google-Autofuzz , @rwhitworth ,

See: https://github.com/uclouvain/openjpeg/pull/882

opj_decompress -i poc-752c5d6632455bac7732b47da0ba98eacf4cfd9bd98b53f5a703504a8ffb6cd5.j2k -o test.png

[INFO] Start to read j2k main header (0).
[ERROR] Invalid precision and/or sgnd values for comp[1]:
        [0] prec(12) sgnd(0) [1] prec(2) sgnd(0)
[ERROR] Marker handler function failed to read the marker segment
ERROR -> opj_decompress: failed to read the header

**Comment by Google-Autofuzz** [comment]
Hi,

Sorry for the delay.

It does look like this is likely a dupe of #855.  The PR mentioned above looks to still be open, is that planned to be accepted?

If there's anything we can do to be of help, please let us know.

 Google AutoFuzz Team

**Comment by kcc** [comment]
This (os something very similar) is  also reported as https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=2532

**Comment by rouault** [comment]
Fixed per d27ccf01c68a31ad62b33d2dc1ba2bb1eeaafe7b
